### PR TITLE
[MODGOBI-132]-updated VendorDetails Mapping

### PR DESCRIPTION
**Pull request by SinghAdes**
## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/gobi/Mapper.java [hunk 1]
```diff
@@ -823,9 +825,26 @@ private void mapVendorDetail(List<CompletableFuture<?>> futures, VendorDetail ve
...
     });
 
-    Optional.ofNullable(mappings.get(Mapping.Field.VENDOR_ACCOUNT))
+    //[MODGOBI-132]
+    //match subAccount send to GOBI with the accountNo established in Organisation, if match found then map otherwise map  the received subaccount as accountNo.
+    Optional.ofNullable(mappings.get(Mapping.Field.VENDOR))
       .ifPresent(field -> futures.add(field.resolve(doc)
-        .thenAccept(o -> vendorDetail.setVendorAccount((String) o))
+        .thenAccept(o -> {
+          if (o != null) {
```
Comment (AndreiMPV) [inline src/main/java/org/folio/gobi/Mapper.java:833]: Please move lines 833 to 844 it to private method

File: src/main/java/org/folio/gobi/Mapper.java [hunk 2]
```diff
@@ -823,9 +825,26 @@ private void mapVendorDetail(List<CompletableFuture<?>> futures, VendorDetail ve
...
+          Organization organization = (Organization) o;
+          Optional.ofNullable(mappings.get(Mapping.Field.VENDOR_ACCOUNT))
+          .ifPresent(vendorAccountfield -> {
+              if (!HelperUtils.extractSubAccount(organization.getAccounts().get(0).getAccountNo()).equals(vendorAccountfield.resolve(doc).join())) {
```
Comment (AndreiMPV) [inline src/main/java/org/folio/gobi/Mapper.java:837]: NPE can be possible organization.getAccounts().get(0).getAccountNo()

Comment (AndreiMPV) [reply src/main/java/org/folio/gobi/Mapper.java:837]: Also don't include int equals complex logic lie vendorAccountfield.resolve(doc).join()
if it fail then will be difficult to understand a root cause

**Review by AndreiMPV — CHANGES_REQUESTED** [review]

File: src/main/java/org/folio/gobi/Mapper.java [hunk 3]
```diff
@@ -823,12 +823,35 @@ private void mapVendorDetail(List<CompletableFuture<?>> futures, VendorDetail ve
...
         .exceptionally(Mapper::logException)));
   }
 
+  private void mapVendorAccount(List<CompletableFuture<?>> futures, Organization organization, VendorDetail vendorDetail, Document doc)
+  {
+    Optional.ofNullable(mappings.get(Mapping.Field.VENDOR_ACCOUNT))
+      .ifPresent(vendorAccountField -> {
+        String vendorAccountNo = (String) vendorAccountField.resolve(doc).join();
+        Optional.ofNullable(organization.getAccounts().get(0).getAccountNo())
+          .ifPresentOrElse(organizationAccountNo ->{
```
Comment (AndreiMPV) [inline src/main/java/org/folio/gobi/Mapper.java:843]: Organization can have or can't have account and line "organization.getAccounts().get(0)" in that case will throw NPE

**Review by AndreiMPV — CHANGES_REQUESTED** [review]

**Comment by grabsefx** [comment]
Please make this mapping configurable by adding lookup to the configuration files https://github.com/folio-org/mod-gobi/tree/master/src/main/resources

**Comment by SinghAdes** [comment]
> Please make this mapping configurable by adding lookup to the configuration files https://github.com/folio-org/mod-gobi/tree/master/src/main/resources

@grabsefx :- I may not be sure about the flow but i think we are already doing a lookup here  :-
[https://github.com/folio-org/mod-gobi/pull/162/files/93fc2a88b5b394ab3831f17a177d67e74989f75a#:~:text=organization%20%3[…]ccountfield%20%2D%3E%20%7B](https://github.com/folio-org/mod-gobi/pull/162/files/93fc2a88b5b394ab3831f17a177d67e74989f75a#:~:text=organization%20%3D%20(Organization)%20o%3B-,Optional.ofNullable(mappings.get(Mapping.Field.VENDOR_ACCOUNT)),-.ifPresent(vendorAccountfield%20%2D%3E%20%7B)

i would be thankful if you can help me on this . thanks in advance.

**Review by grabsefx — APPROVED** [review]

**Comment by grabsefx** [comment]
> > Please make this mapping configurable by adding lookup to the configuration files https://github.com/folio-org/mod-gobi/tree/master/src/main/resources
> 
> @grabsefx :- I may not be sure about the flow but i think we are already doing a lookup here :- [https://github.com/[/pull/162](https://github.com/folio-org/mod-gobi/pull/162)/files/93fc2a88b5b394ab3831f17a177d67e74989f75a#:~:text=organization%20%3[…]ccountfield%20%2D%3E%20%7B](https://github.com/folio-org/mod-gobi/pull/162/files/93fc2a88b5b394ab3831f17a177d67e74989f75a#:~:text=organization%20%3D%20(Organization)%20o%3B-,Optional.ofNullable(mappings.get(Mapping.Field.VENDOR_ACCOUNT)),-.ifPresent(vendorAccountfield%20%2D%3E%20%7B)
> 

ok. leave it as is then. approved

**Review by AndreiMPV — APPROVED** [review]

**Pull request merged**